File: src/core/expr/code/render.rb

module Render
  module RenderExpr
    
    include Interpreter::Dispatcher    
      
    def render(obj)
      dispatch_obj(:render, obj)
    end
  
    def render_EBinOp(obj)
      "#{render(obj.e1)} #{obj.op} #{render(obj.e2)}"
    end
  
    def render_EUnOp(obj)
      "#{obj.op} #{render(obj.e)}"
    end
  
    def render_EField(obj)
      "#{render(obj.e)}.#{obj.fname}"
    end
  
    def render_EVar(obj)
      "#{obj.name}"
    end
  
    def render_EConst(obj)
      "#{obj.val}"
    end
    
    def render_ENil
      ""
    end
  end
  
  class RenderExprC
    include RenderExpr
  end

  def self.eval(obj, args={})
    interp = EvalExprC.new
    interp.dynamic_bind args do
      interp.eval(obj)
    end
  end
end

File: src/js/core/expr/code/render.js
define([
  "core/semantics/code/interpreter"
],
function(Interpreter) {
  var Render ;
  var RenderExpr = MakeMixin([Interpreter.Dispatcher], function() {
    this.render = function(obj) {
      var self = this; 
      return self.dispatch_obj("render", obj);
    };

    this.render_EBinOp = function(obj) {
      var self = this; 
      return S(self.render(obj.e1()), " ", obj.op(), " ", self.render(obj.e2()));
    };

    this.render_EUnOp = function(obj) {
      var self = this; 
      return S(obj.op(), " ", self.render(obj.e()));
    };

    this.render_EField = function(obj) {
      var self = this; 
      return S(self.render(obj.e()), ".", obj.fname());
    };

    this.render_EVar = function(obj) {
      var self = this; 
      return obj.name();
    };

    this.render_EConst = function(obj) {
      var self = this; 
      return obj.val();
    };

    this.render_ENil = function() {
      var self = this; 
      return "";
    };
  });

  var RenderExprC = MakeClass("RenderExprC", null, [RenderExpr],
    function() {
    },
    function(super$) {
    });

  Render = {
    RenderExpr: RenderExpr,
    RenderExprC: RenderExprC,

  };
  return Render;
})
